docs/lesson-plans/week-5.md

```diff
@@ -0,0 +1,639 @@
+
+You can see all of the sources of parameterized arguments [on the JUnit 
+docs](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests-sources).
+
+Here is an example of a parameterized test using integer inputs:
+
+```java
+@ParameterizedTest
+@ValueSource(ints = {1, 2, 3})
+@NullSource
+// Notice how we're using the wrapped integer class "Integer" since we're inputting a null source.
+public void testParameterizedTest(Integer value) { 
+    if (value == null) {
+        assertNull(value);
+    }
+    else {
+        assertTrue(value > 0 && value < 4);
+    }
+}
+```
+
+##### Externals Class
+
+Sometimes you end up testing things where you need to supply your own value (especially if the method providing a value 
+is static). For example, when getting a `Router` object in `PostsRouter`. If we have our own version of a `Router` 
+object which we want to supply for testing purposes, we need a way to substitute that in.
+
+An `Externals` class (sometimes abbreviated to `externs`) can allow us to do that. We can define a 'nested'
+`public static class Externals` as a nested class in `PostsRouter`, which has a method 
+`public Router getRouter(Vertx vertx)`, and that will allow us to do so. In the default `Externals` 
+(the one in `PostsRouter`),  we'll start out just having that single `getRouter` method. Then, in our test class,
+we'll extend that `Externals` class with a new `TestExternals` class, and override the `getRouter` method to have it 
+return our custom router object. This can be applied in many different ways for times where you need to override 
+behavior in dependencies needed for testing.
+
+>A nested class is a class that is created inside of another one. In general, having separate classes can be better, but
+>some reasons to nest classes are because:
+>
+>- It's a way of logically grouping related classes together
+>- The nested class is allowed access to the outer class' private fields/methods
+>- The nested class can be private (available to only the outer class) 
+>- It can be more readable and maintainable 
+>
+>Here's an example of one
+>```java 
+>public class OuterClass {
+>  public class NestedClass { }
+>}
+>```
+
+>A **_static_** nested class is a nested class which doesn't depend on an instance of the outer class to be created.
+>Usually, when creating a nested class, you need to create/have an outer class to create the nested class from, but
+>static nested classes bypass that. However, if a nested class is declared static, then it can't access any non-static
+>methods of its outer class.
+>
+>For example, if we have the `OuterClass` and `NestedClass` above, and we added the following `StaticNestedClass` to 
+>the outer class, here is how we would access them.
+>
+>The StaticNestedClass.
+>
+>```java 
+>... // in OuterClass
+>  public static class StaticNestedClass { }
+>```
+>
+>How to get instances of them.
+>
+>```java 
+>// Non-static
+>OuterClass outerClass = new OuterClass();
```

Comment:
Name the instances something different like `outerInstance` so that people don't get confused by the capitalization